Modules/System Tests/AL Rest Client/src/ALRestClientTests.Codeunit.al

```diff
@@ -0,0 +1,395 @@
+        Assert.AreEqual(GetJsonToken(JsonObject, 'url').AsValue().AsText(), 'https://httpbin.org/put', 'The response should contain the expected url');
+        Assert.AreEqual(GetJsonToken(JsonObject, 'data').AsValue().AsText(), 'Hello World', 'The response should contain the expected data');
+    end;
+
+    [Test]
+    procedure TestDelete()
+    var
+        ALRestClient: Codeunit "AL Rest Client";
+        ALHttpResponseMessage: Codeunit "AL Http Response Message";
+        JsonObject: JsonObject;
+    begin
+        // [SCENARIO] Test DELETE request
+
+        // [GIVEN] An uninitialized AL Rest Client
+        // [WHEN] The Delete method is called
+        ALHttpResponseMessage := ALRestClient.Delete('https://httpbin.org/delete');
+
+        // [THEN] The response contains the expected data
+        Assert.AreEqual(ALHttpResponseMessage.HttpStatusCode(), 200, 'The response status code should be 200');
+        Assert.IsTrue(ALHttpResponseMessage.IsSuccessStatusCode(), 'IsSuccessStatusCode should be true');
+        JsonObject := ALHttpResponseMessage.Content().AsJson().AsObject();
+        Assert.AreEqual(GetJsonToken(JsonObject, 'url').AsValue().AsText(), 'https://httpbin.org/delete', 'The response should contain the expected url');
+    end;
+
+    [Test]
+    procedure TestGetWithDefaultHeaders()
+    var
+        ALRestClient: Codeunit "AL Rest Client";
+        ALHttpResponseMessage: Codeunit "AL Http Response Message";
+        JsonObject: JsonObject;
+    begin
+        // [SCENARIO] Test GET request with headers
+
+        // [GIVEN] An initialized AL Rest Client with default request headers
+        ALRestClient.SetDefaultRequestHeader('X-Test-Header', 'Test');
+
+        // [WHEN] The Get method is called with headers
+        ALHttpResponseMessage := ALRestClient.Get('https://httpbin.org/get');
+
+        // [THEN] The response contains the expected data
+        Assert.AreEqual(ALHttpResponseMessage.HttpStatusCode(), 200, 'The response status code should be 200');
+        Assert.IsTrue(ALHttpResponseMessage.IsSuccessStatusCode(), 'IsSuccessStatusCode should be true');
+        JsonObject := ALHttpResponseMessage.Content().AsJson().AsObject();
+        Assert.AreEqual(GetJsonToken(JsonObject, 'url').AsValue().AsText(), 'https://httpbin.org/get', 'The response should contain the expected url');
+        Assert.AreEqual(SelectJsonToken('$.headers.X-Test-Header', JsonObject).AsValue().AsText(), 'Test', 'The response should contain the expected header');
+    end;
+
+    [Test]
+    procedure TestBaseAddress()
+    var
+        ALRestClient: Codeunit "AL Rest Client";
+        ALHttpResponseMessage: Codeunit "AL Http Response Message";
+        JsonObject: JsonObject;
+    begin
+        // [SCENARIO] Test GET request with base address
+
+        // [GIVEN] An initialized AL Rest Client with base address
+        ALRestClient.SetBaseAddress('https://httpbin.org');
```

Comment:
This tests `'https://httpbin.org' + '/get'` (the happy-path), but what about cases like:
`'https://httpbin.org/' + 'get'`
`'https://httpbin.org' + 'get'`
`'https://httpbin.org/' + '/get'`
Should the module help and make sure the url ends up having a single slash in the correct place?
Or at least in the summary comment?
Alternatively, why expose the concept of a BaseAddress and relative urls, instead of just always requiring a full/absolute url?

Reply:
First of all, this is exposed because the HttpClient features this.
Unfortunately, the HttpRequestMessage in AL does not support relative URLs. While the DotNet object does support it, so this is simply a overlooked feature.
But for this reason, I combine both base address and the request uri when constructing the HttpRequestMessage. Because I don't know how somebody wants to use it, I just combine the two. It's up to the developer to provide correct values, I'm not going to babysit them. 😊

The option of using a base address can be very useful in case the Rest Client is used in a codeunit that supports multiple API endpoints of a server. The Rest Client can be initialized one time with the base address and then the functions can use the relative URL. This is actually the best way of using the Http Client anyway.